Fix multiple issues with lldb 15-17

Itaros opened the pull request:
## Description of changes

 - Fixes #232689
 - Fixes #244609
 - Fixes #166604

Not really proud of 9ccc2f36150a1951c2a6126c0638a05dde913fd3 . Very hacky, but I can't figure out why rpath entry become missing after introduction of 251c55700d3ef977506c3f664888780f707f4a55 , and that is the only way I could do it without hindering any potential improvements. I guess working, but hacky is better then pretending to be working and failing, at least for a while.

```
[nix-shell:~/.cache/nixpkgs-review/rev-f4be6f8b21d7e3552a464294121f6ae2c0fde466]$ cat report.json 
{
    "blacklisted": [],
    "broken": [],
    "built": [
        "lldb_15",
        "lldb_15.dev",
        "lldb_15.lib",
        "lldb_16",
        "lldb_16.dev",
        "lldb_16.lib",
        "llvmPackages_15.lldb-manpages",
        "llvmPackages_16.lldb-manpages"
    ],
    "extra-nixpkgs-config": null,
    "failed": [],
    "non-existant": [],
    "pr": null,
    "system": "x86_64-linux",
    "tests": []
}

```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Itaros committed:
llvmPackages_{14,15,16,17,git}.lldb: unify lua and python path patch

Artturin commented on pkgs/development/compilers/llvm/common/lldb.nix:
```diff
@@ -160,6 +160,9 @@ stdenv.mkDerivation (rec {
   '';
 
+  # TODO: Hack to make dynamic linking work again. Find a better way to do this using the build-system machinery
+  postFixup = if (lib.versions.major release_version) == "16" then ''patchelf --add-rpath ${libclang.lib}/lib $out/bin/.lldb-wrapped'' else null;
```
```suggestion
  postFixup = lib.optionalString (lib.versionAtLeast release_version "15") "patchelf --add-rpath ${libclang.lib}/lib $out/bin/.lldb-wrapped";
```

Ninja3047 commented on pkgs/development/compilers/llvm/common/lldb.nix:
```diff
@@ -160,6 +160,9 @@ stdenv.mkDerivation (rec {
   '';
 
+  # TODO: Hack to make dynamic linking work again. Find a better way to do this using the build-system machinery
+  postFixup = if (stdenv.isLinux && lib.versionAtLeast release_version "16") then "patchelf --add-rpath ${libclang.lib}/lib $out/bin/.lldb-wrapped" else null;
```
this needs to also patch `lldb-server`, `lldb-instr`, `lldb-argdumper`, `lldb-vscode` and potentially others

Jackaed commented:
Any reason why this hasn't been merged yet?

hlolli commented:
I made a pull-request on your branch for a fix, that at least fixes the broken symlink on darwin, I have a feeling it's an issue on linux as well, but I'll need someone else to confirm it.

https://github.com/Itaros/nixpkgs/pull/1

Not urgent that you include this change, I can make another PR after you've had this one approved

hlolli commented:
it seems the paths keep being wrong, not only are the symlinks pointing to nowhere, but I think also that lldb is looking in the wrong place on osx under nix

_lldb.cpython-311-darwin.so -> ../../../liblldb.dylib

is always incorrect for me, no matter how I try relinking this or copying, the multiple outputs may be throwing me off

I have been able to get this fixed

lldb-argdumper -> ../../../../../../../tmp/nix-build-lldb-14.0.6.drv-0/lldb-src-14.0.6/lldb/build/bin/lldb-argdumper

but it doesn't seem to fix the underlying issue, I'm a bit stuck and I hope someone else takes a look at this.

https://github.com/llvm/llvm-project/blob/6266b964202336a02f40007928719e060bc81694/lldb/source/Host/common/HostInfoBase.cpp#L248

I feel it's giving us a clue that it should be finding this under Framework.LLDB but I haven't seen any framework outputs when I build lldb on darwin+nix. Maybe a patch will fix it for darwin users. I'm not sure if there's a better solution or it's already being worked on elsewhere.

wegank commented:
`_lldb.cpython-311-darwin.so -> ../../../liblldb.dylib` is a working relative symlink.

`lldb-argdumper -> ../../../../bin/lldb-argdumper` is not, however, but it is broken for all version of LLDB in nixpkgs. I'll have a look.

Nanotwerp commented:
For the sake of all things good and sacred, we *HAVE* to get this merged!

In the meantime, this fixed LLDB can be applied by using a flake, although you'll have to compile LLDB during the first build:

```
{
  description = "Assembly development environment";

  inputs = {
    nixpkgs.url = "nixpkgs/nixos-unstable";
    lldb-fix.url = "github:Itaros/nixpkgs/sideport-lldb-1x";
  };

  outputs = { self, nixpkgs, lldb-fix }:
    let
      allSystems = [ "x86_64-linux" "aarch64-linux" "x86_64-darwin" "aarch64-darwin" ];
    
      forAllSystems = f: nixpkgs.lib.genAttrs allSystems (system: f {
        pkgs = import nixpkgs { inherit system; };
        lldb-pkgs = import lldb-fix { inherit system; };
      });
    in { 
      devShell = forAllSystems
        ({ pkgs, lldb-pkgs }:
          pkgs.mkShell {
            packages = [
              pkgs.yasm
              pkgs.mold
              lldb-pkgs.lldb_17 
            ]; 
        }); 
    };
}
```


This can then be used with the `nix develop` command, or the better alternative of [nix-direnv](https://github.com/nix-community/nix-direnv), to temporarily enter an environment where all things are well in the world.

sternenseemann commented on pkgs/development/compilers/llvm/common/lldb.nix:
```diff
@@ -154,6 +154,13 @@ stdenv.mkDerivation (rec {
   '';
 
+  # TODO: Hack to make dynamic linking work again. Find a better way to do this using the build-system machinery
+  postFixup = if (stdenv.isLinux && lib.versionAtLeast release_version "16") then ''
```
I guess this is the biggest problem—such hacky workarounds have a tendency to stick which is annoying. Can you tell what changed so this broke between lldb 15 and lldb 16?

Itaros replied:
As I noted in the PR description - I can't figure it out: whole llvm build procedure is quite... enigmatic. I understand the gravity of the hack. My reasoning is: if there is no one to figure it out in a reasonable timeframe it should be accepted as is unless there are potential issues which are unrelated and may result in fallout far bigger than completely broken LLDB.

sternenseemann reviewed (commented):
Seems good except for the patchelfing, the updated patch would also need to b ported to llvmPackages_git, so future updates will benefit from it.

patryk4815 commented:
@Itaros Can we add passthru.tests for this bugs?

My patch:
```patch
diff --git a/pkgs/development/compilers/llvm/common/lldb.nix b/pkgs/development/compilers/llvm/common/lldb.nix
index 2444c1795a78..82ce2bf0e953 100644
--- a/pkgs/development/compilers/llvm/common/lldb.nix
+++ b/pkgs/development/compilers/llvm/common/lldb.nix
@@ -36,7 +36,7 @@ let
       '' else src;
 in

-stdenv.mkDerivation (rec {
+stdenv.mkDerivation (finalAttrs: {
   passthru.monorepoSrc = monorepoSrc;
   pname = "lldb";
   inherit version;
@@ -46,7 +46,7 @@ stdenv.mkDerivation (rec {

   outputs = [ "out" "lib" "dev" ];

-  sourceRoot = lib.optional (lib.versionAtLeast release_version "13") "${src.name}/${pname}";
+  sourceRoot = lib.optional (lib.versionAtLeast release_version "13") "${finalAttrs.src.name}/${finalAttrs.pname}";

   nativeBuildInputs = [
     cmake
@@ -101,7 +101,7 @@ stdenv.mkDerivation (rec {
   hardeningDisable = [ "format" ];

   cmakeFlags = [
-    "-DLLDB_INCLUDE_TESTS=${if doCheck then "YES" else "NO"}"
+    "-DLLDB_INCLUDE_TESTS=${if finalAttrs.doCheck then "YES" else "NO"}"
     "-DLLVM_ENABLE_RTTI=OFF"
     "-DClang_DIR=${lib.getDev libclang}/lib/cmake"
     "-DLLVM_EXTERNAL_LIT=${lit}/bin/lit"
@@ -122,27 +122,32 @@ stdenv.mkDerivation (rec {
     #
     # so, we just ignore the resulting errors
     "-DSPHINX_WARNINGS_AS_ERRORS=OFF"
-  ]) ++ lib.optionals doCheck [
+  ]) ++ lib.optionals finalAttrs.doCheck [
     "-DLLDB_TEST_C_COMPILER=${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc"
     "-DLLDB_TEST_CXX_COMPILER=${stdenv.cc}/bin/${stdenv.cc.targetPrefix}c++"
   ];

   doCheck = false;
-  doInstallCheck = lib.versionOlder release_version "15";
-
-  # TODO: cleanup with mass-rebuild
-  installCheckPhase = ''
-    if [ ! -e $lib/${python3.sitePackages}/lldb/_lldb*.so ] ; then
-        echo "ERROR: python files not installed where expected!";
-        return 1;
+
+  # check if lldb works fine, this skip symlinks issues
+  passthru.tests.lldb_check = runCommand "${finalAttrs.pname}-test" {} (''
+    broken=0
+    ${finalAttrs.finalPackage}/bin/lldb --script-language python -o 'script print("python", None)' -o 'quit' > test_py
+    if ! grep -q 'python None' ./test_py; then
+      echo "Error: broken lldb python interpreter"
+      broken=1
     fi
-  '' # Something lua is built on older versions but this file doesn't exist.
-  + lib.optionalString (lib.versionAtLeast release_version "14") ''
-    if [ ! -e "$lib/lib/lua/${lua5_3.luaversion}/lldb.so" ] ; then
-        echo "ERROR: lua files not installed where expected!";
-        return 1;
+  '' + (lib.optionalString (lib.versionAtLeast release_version "10") ''
+    ${finalAttrs.finalPackage}/bin/lldb --script-language lua -o 'script print("lua", nil)' -o 'quit' > test_lua
+    if ! grep -qP 'lua\tnil' ./test_lua; then
+      echo "Error: broken lldb lua interpreter"
+      broken=1
     fi
-  '';
+  '') + ''
+    if [ $broken -eq 0 ]; then
+      mkdir $out
+    fi
+  '');

   postInstall = ''
     wrapProgram $out/bin/lldb --prefix PYTHONPATH : $lib/${python3.sitePackages}/
```

sternenseemann committed:
llvmPackages_{16,17,git}.lldb: tell cc-wrapper about libclang*.so

It appears that LLVM's build system no longer sets the executable's
rpath to include the faux resource root we pass in, so we need to make
sure cc-wrapper does this.

sternenseemann reviewed (approved):
I've pushed an improved solution for the rpath problem. What remains to be done is testing the builds (which I've kicked off) and updating the patch for llvmPackages_git which ought not to be forgotten.

sternenseemann reviewed (changes requested):
I've pushed an improved solution for the rpath problem. What remains to be done is testing the builds (which I've kicked off) and updating the patch for llvmPackages_git which ought not to be forgotten.

Nanotwerp commented:
> I've pushed an improved solution for the rpath problem. What remains to be done is testing the builds (which I've kicked off) and updating the patch for llvmPackages_git which ought not to be forgotten.

Tested in a nix shell, I can confirm that these patches work for lldb and all of its subcommands.

sternenseemann reviewed (approved):
llvmPackages_git failure is not a regression: #278667.
